### PR TITLE
Types for compiled template modules

### LINKED ISSUE #242: [discussion] How should we handle templates?
Currently, importing a template (i.e., in an addon's component) includes a 

##### lib/my-addon/addon/components/x-foo.ts
```ts
// @ts-ignore: Ignore import of compiled template
import layout from '../templates/components/x-foo';
```
to get around the issue of typescript being unable to resolve the module. Since we're already maintaining a lot of bookkeeping for addons (in particular, `tsconfig.json` paths/includes for in-repo addons/engines), have we considered 

##### lib/my-addon/addon/components/x-foo.ts
```ts
import layout from 'my-addon/templates/components/x-foo';
```
along with
##### (A) types/my-addon.d.ts
```ts
declare module 'my-addon/templates/*' {
  import { TemplateFactory } from "htmlbars-inline-precompile";
  export default TemplateFactory
}
```
or 

##### (B) types/my-addon.d.ts
```ts
declare module 'my-addon/templates/*' {
  export default any
}
```

or 

##### (C) types/my-addon.d.ts
```ts
declare module 'my-addon/templates/*' {
  export default any
}
```

### PR DESCRIPTION
**Pull request by mike-north**
Deliberately based off of #644, which is based off of #647
Fixes #242 
Related #192

This PR results in a `types/global.d.ts` file being generated, producing a module type augmentation for paths we expect to contain templates

```ts
  import { TemplateFactory } from 'htmlbars-inline-precompile';
  const tmpl: TemplateFactory;
  export default tmpl;
```

Depending on project layout the path of the module will be different

*Classic*
```ts
declare module 'foo/templates/*' { ... }
```
*Pods*
```ts
declare module 'foo/components/*/template' { ... }
```
*Module Unification*
```ts
declare module 'foo/ui/components/*/template' { ... }
```

### CODE REVIEW

File: ts/blueprints/ember-cli-typescript/index.js [hunk 1]
```diff
@@ -3,16 +3,36 @@
 const fs = require('fs');
 const path = require('path');
 
-const APP_DECLARATIONS = `
-import Ember from 'ember';
+const APP_DECLARATIONS = `import Ember from 'ember';
```
Comment (mike-north) [inline ts/blueprints/ember-cli-typescript/index.js]: unrelated fix to avoid an unnecessary newline before the `types/my-app/index.d.ts` imports

File: ts/blueprints/ember-cli-typescript/index.js [hunk 2]
```diff
@@ -3,16 +3,36 @@
...
+  const moduleBody = `
+  import { TemplateFactory } from 'htmlbars-inline-precompile';
+  const tmpl: TemplateFactory;
+  export default tmpl;
 `;
+  switch (layout) {
+    case 'classic': return `${comment}
+declare module '${projectName}/templates/*' { ${moduleBody}}`;
+    case 'pods': return `${comment}
+declare module '${projectName}/components/*/template' { ${moduleBody}}`;
```
Comment (dfreeman) [inline ts/blueprints/ember-cli-typescript/index.js]: I think this needs to account for `podModulePrefix`—it will also miss any route templates that are defined. Given how loose things are with pods, the simplest approach might just be `declare module '${projectName}/*/template' {`

File: ts/blueprints/ember-cli-typescript/index.js [hunk 3]
```diff
@@ -3,16 +3,36 @@
...
+    case 'mu': return `${comment}
+declare module '${projectName}/ui/components/*/template' { ${moduleBody}}`;
+    default: throw new Error(`Unexpecte project layout type: "${layout}"`);
```
Comment (dfreeman) [inline ts/blueprints/ember-cli-typescript/index.js]: Typo

```suggestion
    default: throw new Error(`Unexpected project layout type: "${layout}"`);
```

File: ts/blueprints/ember-cli-typescript/index.js [hunk 4]
```diff
@@ -33,6 +53,7 @@ module.exports = {
     let isAddon = this.project.isEmberCLIAddon();
     let isMU = this._detectMU();
     let includes = isMU ? ['src'] : ['app', isAddon && 'addon'].filter(Boolean);
+    const isPods = this.pod;
```
Comment (dfreeman) [inline ts/blueprints/ember-cli-typescript/index.js]: Since the default blueprint is normally run as part of `ember install`, I believe this flag will only ever be set if we have users explicitly re-run the generator afterwards with `ember g ember-cli-typescript --pod`. Is that something we want to encourage?

Comment (mike-north) [reply ts/blueprints/ember-cli-typescript/index.js]: I tested this manually with the `.embercli.js` file having the pods flag set to true. The scenario I'm thinking of is


```shell
ember new i-should-not-use-pods --pods # sets up .embercli.js with "pods": true
cd i-should-not-use-pods
ember install ember-cli-typescript

```

File: ts/blueprints/ember-cli-typescript/index.js (hunk 2 above)
Comment (mike-north) [reply ts/blueprints/ember-cli-typescript/index.js]: Thanks! I'll go with your "simplest approach" idea

File: ts/blueprints/ember-cli-typescript/index.js (hunk 4 above)
Comment (dfreeman) [reply ts/blueprints/ember-cli-typescript/index.js]: Gotcha; I'd forgotten you could set a default there
